src/pveclib/vec_int64_ppc.h

```diff
@@ -3584,6 +3768,154 @@ vec_xxspltd (vui64_t vra, const int ctl)
   return (result);
 }
 
+/** \brief Vector Load Scalar doubleword indexed.
+ *
+ *  This operation is an alternate form of vector load element, with
+ *  the added simplification that data is always left justified in the
+ *  vector. This simplifies merging elements for gather operations.
+ *
+ *  \note This is instruction was introduced in PowerISA 2.06 (POWER7).
+ *  For POWER8/9 there are additional optimizations by effectively
+ *  converting small constant index values into displacements. For
+ *  POWER8 a specific pattern of addi/lsxdx instruction is <I>fused</I>
+ *  into a single load displacement internal operation. For POWER9 we can
+ *  use the lxsd (DS-form) instruction directly.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   |   5   | 2/cycle  |
+ *  |power9   |   5   | 2/cycle  |
+ *
+ *  @param ra const doubleword index (displacement).
+ *  @param rb const doubleword pointer to an array of doublewords.
+ *  @return The data (ra + rb) is loaded into vector doubleword
+ *  element 0. Element 1 is undefined.
+ */
+static inline vui64_t
+vec_vlxsdx (const unsigned long long ra, const unsigned long long *rb)
+{
+  vui64_t xt;
+
+  if (__builtin_constant_p (ra) && (ra < 32760) && ((ra & 3) == 0))
+    {
+#if defined (_ARCH_PWR9)
+      __asm__(
+	  "lxsd%X1 %0,%1;"
+	  : "=v" (xt)
+	  : "m" (*((char *)rb + ra))
+	  : );
+#else
+      if (ra == 0)
+	{
+	  __asm__(
+	      "lxsdx %x0,%y1;"
+	      : "=wa" (xt)
+	      : "Z" (*rb)
+	      : );
+	} else {
+	  // Would be better if li and lxsdx shared a single asm block
+	  // (enforcing consecutive instructions).
+	  // This enables instruction fusion for P8.
+	  unsigned long long rt;
+#if 0     // Expected this combination of constraints to work. But the
+	  // compiler does not recognize the early clobber of rt and
+	  // generates bad code.
+	  __asm__(
+	      "li %0,%2;"
+	      "lxsdx %x1,%y3;"
+	      : "=&r" (rt), "=wa" (xt)
+	      : "I" (ra), "Z" (*((char *)rb+rt))
+	      : );
+#else
+	  // This generates operationally correct code, but the compiler
+	  // may rearrange code and break the fusion pattern.
+	  __asm__(
+	      "li %0,%1;"
+	      : "=r" (rt)
+	      : "I" (ra)
+	      : );
+	  __asm__(
+	      "lxsdx %x0,%y1;"
+	      : "=wa" (xt)
+	      : "Z" (*((char *)rb+rt))
+	      : );
+#endif
+	}
+#endif
+    } else {
+      __asm__(
+	  "lxsdx %x0,%y1;"
+	  : "=wa" (xt)
+	  : "Z" (*((char *)rb+ra))
+	  : );
+    }
+  return xt;
+}
+
+/** \brief Vector Store Scalar doubleword indexed.
+ *
+ *  This operation is an alternate form of vector store element, with
```

Comment:
same as previous: document what this actually does.

Reply:
Store the left most doubleword of vector xs as a scalar doubleword
at the effective address formed by rb+ra.